plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java

```diff
@@ -0,0 +1,202 @@
+        if (response instanceof PayloadResponse && ((PayloadResponse) response).getPayload() instanceof Content) {
+          metadataContents.put(entry.getKey(), (Content) ((PayloadResponse) response).getPayload());
+        }
+      }
+    }
+
+    if (metadataContents.isEmpty()) {
+      return null;
+    }
+    return doMergeAndCacheMetadata(mavenPath, metadataContents);
+  }
+
+  private Content doMergeAndCacheMetadata(final MavenPath mavenPath,
+                                          final LinkedHashMap<Repository, Content> metadataContents) throws IOException
+  {
+    if (metadataContents.size() == 1) {
+      return metadataContents.get(0);
+    }
+    final MetadataXpp3Reader reader = new MetadataXpp3Reader();
+    List<MetadataEnvelope> metadatas = Lists.newArrayList();
+    for (Map.Entry<Repository, Content> entry : metadataContents.entrySet()) {
+      final String origin = entry.getKey().getName() + " @ " + mavenPath.getPath();
+      try (InputStream inputStream = entry.getValue().openInputStream()) {
+        final Metadata metadata = reader.read(inputStream);
+        metadatas.add(new MetadataEnvelope(origin, metadata));
+      }
+      catch (XmlPullParserException e) {
+        // skip it, log it
+        log.info("Unparseable repository metadata: {}", origin, e);
+      }
+    }
+
+    final Metadata mergedMetadata = metadataMerger.merge(metadatas.toArray(new MetadataEnvelope[metadatas.size()]));
+    if (mergedMetadata == null) {
+      return null;
+    }
+    final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    new MetadataXpp3Writer().write(byteArrayOutputStream, mergedMetadata);
+    // Metadata is rather small, so let's do it like this
+    final Iterable<HashAlgorithm> hashAlgorithms = Iterables.transform(Arrays.asList(HashType.values()),
+        new Function<HashType, HashAlgorithm>()
+        {
+          @Override
+          public HashAlgorithm apply(final HashType input) {
+            return input.getHashAlgorithm();
+          }
+        });
+    final Map<HashAlgorithm, HashCode> hashCodes = Hashes.hash(hashAlgorithms,
+        new ByteArrayInputStream(byteArrayOutputStream.toByteArray()));
+    final Content content = new Content(
+        new BytesPayload(
+            byteArrayOutputStream.toByteArray(),
+            "text/xml"
+        ));
+    content.getAttributes().set(Content.CONTENT_LAST_MODIFIED, DateTime.now());
+    content.getAttributes().set(Content.CONTENT_ETAG, "{SHA1{" + hashCodes.get(HashAlgorithm.SHA1).toString() + "}}");
+    content.getAttributes().set(Content.CONTENT_HASH_CODES_MAP, hashCodes);
+
+    // cache the metadata and the hashes
+    mavenFacet.put(mavenPath, content);
+    for (HashType hashType : HashType.values()) {
+      final HashCode hashCode = hashCodes.get(hashType.getHashAlgorithm());
+      if (hashCode != null) {
+        mavenFacet.put(mavenPath.hash(hashType), new StringPayload(hashCode.toString(), "text/plain"));
```

Comment:
Personal preference, we should use constants for mime types if possible: https://github.com/sonatype/nexus-oss/blob/master/components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentTypes.java#L20-20